### PR TITLE
Don't recommend `--prerelease=allow` for source dist builds

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

Closes https://github.com/astral-sh/uv/issues/3686.

### CODE REVIEW

**Comment by zanieb** [comment]
I don't entirely follow the motivation in the issue. Can you explain a bit?

**Comment by charliermarsh** [comment]
We don't pass the `--prerelease` flag to source distribution builds. Recommending `--prerelease=allow` has no effect.

**Comment by zanieb** [comment]
Like.. all source distribution builds including those from the registry that are prereleases? Or non-registry ones like local directories or Git repositories?

**Comment by charliermarsh** [comment]
I don't fully follow the distinction you're referencing but we never pass the `--prerelease` flag to source distribution builds. Specifically, when we resolve `build-system.requires`, we don't propagate `--prerelease`, `--resolution`, or `--no-deps`. That applies to _all_ source distribution builds (but _only_ to the `build-system.requires`, like `requires = ["setuptools>=40.8.0"]`).

**Comment by charliermarsh** [comment]
In the linked issue, we failed to resolve `requires = ["setuptools>=40.8.0"]` when building some package. We then recommended passing `--prerelease=allow`, but doing so has no effect.

**Comment by charliermarsh** [comment]
We removed that propagation in https://github.com/astral-sh/uv/pull/1607 if you want to see a few linked issues. I think it's unclear whether `--prerelease` should be propagated (`--resolution` definitely should not, I think that one is much clearer), but it's reasonable for it _not_ to be IMO.

**Comment by zanieb** [comment]
Ah thanks for clarifying. I somehow missed that we're talking about an error when resolving _build requirements_. Hence my extreme confusion and dumb questions. This change makes a ton of sense. Can you rephrase the title to make it a bit more obvious? Like...

> Don't recommend `--prerelease=allow` during build requirement resolution errors

**Comment by charliermarsh** [comment]
Yeah for sure.

**Comment by charliermarsh** [comment]
Thank you!
